src/hotspot/share/opto/superword.cpp

```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
 //------------------------------construct_bb---------------------------
 // Construct reverse postorder list of block members
 bool SuperWord::construct_bb() {
-  Node* entry = bb();
-
   assert(_block.length() == 0,          "block is empty");
-  assert(_data_entry.length() == 0,     "data_entry is empty");
-  assert(_mem_slice_head.length() == 0, "mem_slice_head is empty");
-  assert(_mem_slice_tail.length() == 0, "mem_slice_tail is empty");
 
-  // Find non-control nodes with no inputs from within block,
-  // create a temporary map from node _idx to bb_idx for use
-  // by the visited and post_visited sets,
-  // and count number of nodes in block.
-  int bb_ct = 0;
+  // First pass over loop body:
+  //  (1) Check that there are no unwanted nodes (LoadStore, MergeMem, data Proj).
+  //  (2) Count number of nodes, and create a temporary map (_idx -> bb_idx).
+  //  (3) Verify that all non-ctrl nodes have an input inside the loop.
+  int block_count = 0;
   for (uint i = 0; i < lpt()->_body.size(); i++) {
-    Node *n = lpt()->_body.at(i);
+    Node* n = lpt()->_body.at(i);
     set_bb_idx(n, i); // Create a temporary map
     if (in_bb(n)) {
+      block_count++;
+
       if (n->is_LoadStore() || n->is_MergeMem() ||
           (n->is_Proj() && !n->as_Proj()->is_CFG())) {
         // Bailout if the loop has LoadStore, MergeMem or data Proj
         // nodes. Superword optimization does not work with them.
+#ifndef PRODUCT
+      if (is_trace_superword_any()) {
+        tty->print_cr("SuperWord::construct_bb: fails because of unhandled node:");
+        n->dump();
+      }
+#endif
         return false;
       }
-      bb_ct++;
+
+#ifdef ASSERT
       if (!n->is_CFG()) {
         bool found = false;
         for (uint j = 0; j < n->req(); j++) {
           Node* def = n->in(j);
-          if (def && in_bb(def)) {
+          if (def != nullptr && in_bb(def)) {
             found = true;
             break;
           }
         }
-        if (!found) {
-          assert(n != entry, "can't be entry");
-          _data_entry.push(n);
-        }
```

Comment:
Note: `found` is always true, it turns out. I added an assert. And this means I can also remove `_data_entry`, as we now never push anything to it.